Add regular expression feed filtering

forty8bits opened the pull request:
Mimics Mastodon web's functionality, but in a simpler form; a single regular expression is shared across the home, local and federated feeds. Even uses the exact same labels.

Strings are currently only provided in English and will need to be translated. Not sure what the current state of testing within the app is either, as the unit tests seem to fail for me, so I've included none here for now.

Fixes #555.

charlag commented:
Wow, cool, thanks! How did I miss that? Thank you! I think even in this simple form it may be better than nothing (but @connyduck may disagree).
Don't worry too much about translations, people will submit them as needed.
I'm not sure what happened to tests, will have a look tomorrow.

Welcome to the family! :laughing:

connyduck commented:
You need some kind of error handling. If I enter an invalid regex, Tusky crashes. Just show a toast with something like "invalid regular expression, please try again" :)

forty8bits commented:
Ah, great point, not sure how I overlooked that one. :+1: I'll get that added, although it could be a day or two before I get to it.

forty8bits commented:
@connyduck I got a little time today to take a look. I've added validation and cleaned up some buggy behaviour which was causing crashes. The only thing I can think of which I may be overlooking now is in multi-account functionality; not sure how that works in app, but if it's not just handled automatically by the preference system then I'll need to take some time to create a second account on an instance and test that out, although if either yourself or @charlag have more than one account to test with before I get a chance that'd be great. No worries if not. :+1:

forty8bits commented:
I've just done a quick test, and it seems like both existing tab filtering options (replies and boosts) are shared between any accounts logged into the app. Ideally (I think) these, or at least the regular expression, would (or _could_) be per account, but at least now it seems as though the behaviour of my code is consistent with what's already there.

charlag commented:
I'm sorry that I've been out of loop, too many things!
I would say that common filter is counter-intuitive but I'm not sure if making a separate ones is harder. Do we use account prefixes or?..

forty8bits commented on app/src/main/java/com/keylesspalace/tusky/fragment/PreferencesFragment.java:
```diff
@@ -62,6 +67,31 @@ public void onCreate(Bundle savedInstanceState) {
+                    } catch (IllegalArgumentException e) {
+                        ((AlertDialog) ((EditTextPreference) preference1).getDialog()).getButton(AlertDialog.BUTTON_POSITIVE).setEnabled(false);
+                        ((EditTextPreference) regexPref).getEditText().setError("Invalid regular expression!");
+                    } catch (NullPointerException e) {}
```
Ah, my bad. I did this as a quick workaround when noticing unexpected crashes on exiting the dialog and completely forgot to investigate further and clean up. It's also been a fair few years since I last read Effective Java! :man_shrugging:

forty8bits commented:
> It would be awesome if the regex could be shown as the `summary` of the `EditTextPreference` so users don't have to click it to see what the regex is. (this is a bonus, I would merge it without it as well)

I did a quick test with my usual regex (hardcoded into the template), but it seems like there's no way to truncate it down to one line. Without being able to do this, it can just fill the screen and looks pretty bad! Possibly something for a future UI change away from the simplistic `EditTextPreference` element?

Other than that, I've made the other changes requested. :+1:

connyduck commented:
Looks good, thx for your contribution 😊

connyduck commented:
Looks good, thx for your contribution 😊
